src/test/optimizer/strategy/dependent_group_by_reduction_rule_test.cpp

```diff
@@ -0,0 +1,273 @@
+#include "gtest/gtest.h"
+
+#include "expression/expression_functional.hpp"
+#include "logical_query_plan/delete_node.hpp"
+#include "logical_query_plan/insert_node.hpp"
+#include "logical_query_plan/join_node.hpp"
+#include "logical_query_plan/mock_node.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/projection_node.hpp"
+#include "logical_query_plan/sort_node.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+#include "logical_query_plan/union_node.hpp"
+#include "logical_query_plan/update_node.hpp"
+#include "optimizer/strategy/dependent_group_by_reduction_rule.hpp"
+
+#include "strategy_base_test.hpp"
+#include "testing_assert.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+class DependentGroupByReductionRuleTest : public StrategyBaseTest {
+ public:
+  void SetUp() override {
+    auto& storage_manager = Hyrise::get().storage_manager;
+
+    TableColumnDefinitions column_definitions{
+        {"column0", DataType::Int, false}, {"column1", DataType::Int, false}, {"column2", DataType::Int, false}};
+
+    table_a = std::make_shared<Table>(column_definitions, TableType::Data, 2, UseMvcc::Yes);
+    table_a->add_soft_unique_constraint({ColumnID{0}}, IsPrimaryKey::Yes);
```

Comment:
Have you tested non-primary constraints?

Reply:
To be honest I never cared about them since I only thought about primary keys initially. Is there any reason to not include unique columns the same way we currently do with primary keys?
As far as I see it, the main difference is NULL values which should not matter in the aggregate.

Reply:
Not an answer, but as discussed: Test ANY on completely NULL column

Reply:
Tests still to be done, but as I understand unique constraints and primary keys now, we are fine with any unique constraint. So I would remove the check for primary key. Your opinion?

Reply:
I first thought I would understand NULL, but apparently I don't.

Reply:
So my current understanding: NULL are groupable (which I thought would not make sense, as they would not equal each other). If they are groupable, each single row of a group-by column that has a unique constraint (which allows NULLs) does no longer identify exactly one single row. Hence, we can only use the rule for primary keys.

Your take on that?

Reply:
https://academy.vertabelo.com/blog/null-values-group-clause/

> To explain why NULLs are grouped into one bucket, we need to review the SQL standard. SQL defines “any two values that are equal to one another, or any two NULLs”, as “not distinct”. This definition of “not distinct” allows SQL to group and sort NULLs when the GROUP BY clause (or other keywords that perform grouping) is used.

(1) How would limiting this to primary keys reduce the applicability to the TPC-H queries?
(2) I am not sure how well we track NULLable information. If we can show that there are no NULL values, we are fine, too.

Reply:
> (1) How would limiting this to primary keys reduce the applicability to the TPC-H queries?
> (2) I am not sure how well we track NULLable information. If we can show that there are no NULL values, we are fine, too.

(1) Wouldn't change anything since we don't have unique columns which are not also primary keys. But I don't know if there are further constraints allowed in TPC-H which we might use at a later point in time.
(2) Good point. I will change the check for PK to PK or Unique with Non-Nullable columns. I thought about NULLs that are later on added (e.g., outer joins) but they should not impact the rule as it would be still be safe to remove columns here since they are all NULL anyways for added rows. Does that make sense to you?

Reply:
NULLs that are added later on should result in the expression being NULLable, so I don't see how that is an issue. We are talking about `is_nullable_on_lqp`, not `column_is_nullable`, right?

Reply:
Starting at line 66, I now check if we have a constraint (no matter what kind, thus can be unique constraint or PK) and that this column is not nullable (in that case, unique columns are equal to PK columns from the stand point of this rule).

Reply:
Now, ignoring all group-by columns that are nullable.

Reply:
Further, we now process the constraints by size and continue after one reduction was successful (also tested).